### PR TITLE
Mps nn graph

### PR DESCRIPTION
**Pull request by fujunwei**
fix https://github.com/intel/webml-polyfill/issues/210.

### CODE REVIEW

**Comment by fujunwei** [comment]
@huningxin  PTAL, thanks.

**Comment by fujunwei** [comment]
@huningxin  PTAL  again, thanks.

File: services/ml/compilation_impl_mac.mm [hunk 1]
```diff
@@ -205,5 +124,181 @@
...
+      // mps_image_nodes_[inputs[0]] = temporary_image;
+      next_graph = false;
+    }
+
+    if (!success)
+      break;
+
+    for (size_t i = 0; i < outputs_.size(); i++) {
+      if (outputs[0] == outputs_[i]) {
+        next_graph = true;
```
Comment (huningxin) [inline services/ml/compilation_impl_mac.mm]: Why not use [`MPSNNGraph.initWithDevice:resultImages:resultsAreNeeded:`](https://developer.apple.com/documentation/metalperformanceshaders/mpsnngraph/3037385-initwithdevice?language=objc) for multiple outputs?

Comment (fujunwei) [reply services/ml/compilation_impl_mac.mm]: It will be optimized in 10.14.1+.

File: services/ml/compilation_impl_mac_mps.mm [hunk 2]
```diff
@@ -448,121 +411,125 @@ bool CompileConcatenation(std::vector<OperationMac>& operations,
...
+  for (size_t i = 0; i < inputs.size() - 1; ++i) {
+    uint32_t concat_input_idx = inputs[i];
+    const OperandMac& operand = operands[concat_input_idx];
+    for (size_t j = 0; j < operations.size(); ++j) {
+      OperationMac& operation = operations[j];
+      if (operation.outputs[0] == concat_input_idx) {
+        DLOG(INFO) << "  Rewrite op " << j << " type " << operation.type
+                   << " output from " << operation.outputs[0] << " to "
+                   << concat_output_idx;
+        operation.outputs[0] = concat_output_idx;
```
Comment (huningxin) [inline services/ml/compilation_impl_mac_mps.mm]: do we still need this logic? can we feed inputs directly to `MPSNNConcatenationNode`?

File: services/ml/execution_impl_mac_mps.mm [hunk 3]
```diff
@@ -191,6 +147,19 @@ void API_AVAILABLE(macosx(10.13)) ExecutionImplMacMPS::SetupMPSImageForOperands(
   }
 }
 
+API_AVAILABLE(macosx(10.13))
+void ExecutionImplMacMPS::CreateOutputMTLBuffer() {
+  for (size_t i = 0; i < compilation_->outputs_.size(); ++i) {
+    const OperandMac& operand =
+        compilation_->operands_[compilation_->outputs_[i]];
+    // TODO(junwei), the output_mtlbuffers_ need to be removed instead of output
+    // image size.
```
Comment (huningxin) [inline services/ml/execution_impl_mac_mps.mm]: what does this comment mean?

File: services/ml/execution_impl_mac_mps.mm [hunk 4]
```diff
@@ -218,85 +190,50 @@ void API_AVAILABLE(macosx(10.13)) ExecutionImplMacMPS::SetupMPSImageForOperands(
...
+            source_images = [NSMutableArray arrayWithCapacity:1];
+            // Find temporary input images of next graph.
+            [source_images
+                addObject:temporary_mps_images[compilation_
+                                                   ->temporary_inputs_[i - 1]]];
           }
+          NSMutableArray<MPSImage*>* intermediate_images =
+              [NSMutableArray arrayWithCapacity:1];
+
+          MPSImage* output_image_graph = [compilation_->graphs_[i]
```
Comment (huningxin) [inline services/ml/execution_impl_mac_mps.mm]: `graph_output_image`?

**Comment by huningxin** [comment]
Great work! Finish my review, leave some comments and questions. Please take a look. Thanks!

**Comment by fujunwei** [comment]
@huningxin  PTAL  https://github.com/otcshare/chromium-src/pull/61/commits/8102c9b67f7d8ae62ad2f3ab89935c94f95380ee, thanks.

**Comment by huningxin** [comment]
Looks good. Thanks for the PR! Great work!

**Pull request merged**